src/Microsoft.AspNet.SignalR.Client/Connection.cs

```diff
@@ -519,13 +529,29 @@ Task StartNegotiation()
                                         }
                                         if (!string.IsNullOrEmpty(negotiationResponse.RedirectUrl))
                                         {
-                                            if (!negotiationResponse.RedirectUrl.EndsWith("/"))
+                                            var splitUrlAndQuery = negotiationResponse.RedirectUrl.Split(new[] { '?' }, 2);
+
+                                            // Update the URL based on the redirect response and restart the negotiation
+                                            Url = splitUrlAndQuery[0];
+
+                                            if (splitUrlAndQuery.Length == 2 && !string.IsNullOrEmpty(splitUrlAndQuery[1]))
                                             {
-                                                negotiationResponse.RedirectUrl += "/";
+                                                var sb = new StringBuilder("?");
+                                                UrlBuilder.AppendCustomQueryString(sb, splitUrlAndQuery[1]);
+                                                UrlBuilder.AppendCustomQueryString(sb, _userQueryString);
+
+                                                // Update IConnection.QueryString with query string from only the most recent RedirectUrl.
+                                                _redirectPlusUserQueryString = UrlBuilder.Trim(sb);
```

Comment:
I don't love how this preserves the last redirect query string if the client is restarted, but the same is true for the redirect url...